feature: PresburgerSet subtraction prototype

Dinistro opened the pull request:
This contains the first prototype for the Presburger set subtraction. It is taken from libint but does not support the redundant optimisation.

I'm actually not sure if this works, so I decided to already create a draft pull request to enable discussion on the code. I will continue working on the MLIR integration to then be able to test it.

Dinistro committed:
feature: PresburgerSet subtraction prototype

This contains the first prototype for the Presburger set subtraction. It
is taken from libint but does not support the redundant optimization.

Dinistro commented on mlir/lib/Dialect/Presburger/Set.cpp:
```diff
@@ -72,6 +73,174 @@ void PresburgerSet::intersectSet(const PresburgerSet &set) {
+    // The first inequality is positive and the second is negative, of which
+    // we need the complements (strict negative and strict positive).
+    const auto &eq = S_i.getEquality(j);
+    /*if (isMarkedRedundant[2 * j]) {
```
Do we need this for correctness or is this just an optimisation? In the later case I suggest that I first try to get a working end to end solution that enables testing.

superty replied:
Unfortunately the libint code seems to be incorrect here on master. The NOTs should be inverted here. In fact, I think this heuristic can be omitted until we show that it actually helps noticeably, as it seems to add a good deal of complexity to the code.

I suggest just treating equalities as a pair of inequalities for now, by replacing the body of the loop with:
```cpp
const auto &eq = S_i.getEquality(j);
recurseWithInequalityFromEquality(eq, true, true);
addInequalityFromEquality(eq, false, false);
recurseWithInequalityFromEquality(eq, false, true);
addInequalityFromEquality(eq, true, false);
```

Dinistro replied:
Thanks for the feedback. I will adapt this and we can later on continue with this.

Dinistro committed:
fix: recrusive subtraction now actually recurses

After discussion with Arjun it turned out that the libint implementation
was faulty, therefore we had to change the one here.

Dinistro committed:
feature: subtraction operation prototype

Dinistro commented on mlir/lib/Analysis/Presburger/Simplex.cpp:
```diff
@@ -639,17 +649,18 @@ class GBRSimplex {
 };
 
 // Let b_{level}, b_{level + 1}, ... b_n be the current basis.
-// Let F_i(v) = max <v, x - y> where x and y are points in the original polytope
```
Turns out that my clang-format vim binding doesn't actually change this anymore. Perhaps I accidentally run clang-format with a wrong config at some point. I will fix all the comment changes.

Dinistro committed:
fix: reverted reformated comments

It turned out that clang-format was executed with the wrong config once.

Dinistro committed:
fix: CI change cache id

Dinistro committed:
fix: PresburgerSet subtraction now supports equalies properly

Dinistro committed:
testing: canonicalization tests for set subtraction

Dinistro committed:
chore: small comments for PresburgerSet

Dinistro committed:
feature: PresburgerSet complement

Dinistro committed:
feature: presburger.complement operation + canonicalization

Dinistro committed:
fix: PresburgerSet subtraction added missing inequality to FlatAffineConstraint

Dinistro committed:
fix: PresburgerSet::subtract now again uses a ref

Dinistro committed:
fix: print (1 = 0) for empty set

If a set is marked empty, we better print it as empty. This was a
problem for sets without constraints that were marked empty.

Dinistro committed:
refactoring: add constraint transformation funcs

Dinistro committed:
feature: PresburgerSet integer sampling

Dinistro committed:
feature: check that no symbols are present

As we do currently not support symbols in all cases, we have to assert
that non are present before we try to perform operations.

Dinistro commented:
Closing because theese changes are contained in another branch
